numba/core/runtime/nrt.c

```diff
@@ -212,9 +214,10 @@ void nrt_internal_dtor_safe(void *ptr, size_t size, void *info) {
 
 static
 void *nrt_allocate_meminfo_and_data(size_t size, NRT_MemInfo **mi_out, NRT_ExternalAllocator *allocator) {
-    NRT_MemInfo *mi;
+    NRT_MemInfo *mi = NULL;
     NRT_Debug(nrt_debug_print("nrt_allocate_meminfo_and_data %p\n", allocator));
     char *base = NRT_Allocate_External(sizeof(NRT_MemInfo) + size, allocator);
+    if (base == NULL) return NULL; /* return early if alloc fails */
```

Comment:
I don't think there's a contract defined, but it seems like it'd be a good idea to make this the behaviour so I'll change the code to reflect. I wonder if it'd be good to enforce `mi_out` is `NULL` on input too?

Reply:
Do you mean if the caller passes in a null for `mi_out`? That would cause a leak if memory was successfully allocated, so I think it should panic.

Reply:
I think the general question throughout this code is... what should happen with arguments declared as a pointers that have a detectable and invalid value at runtime. The difficulty is what to do with the "panic", I think the only communication channels available here are to `return NULL;` (or this could be a status value instead, but this would require API changes) or signals (which are probably not a good route given the use cases)? Historically the onus has been on the caller to ensure any calls are made with valid arguments, perhaps this should just remain the case.

Reply:
Braces around returns and setting `*mi_out = NULL;` done in d286f12